### PR TITLE
[p2p/pex] connect to more than 10 peers

### PR DESCRIPTION
**Pull request by melekes**
Refs #2130

* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [ ] Wrote tests
* [x] Updated CHANGELOG_PENDING.md

### CODE REVIEW

**Comment by melekes** [comment]
```go
func TestPEXReactorRunning(t *testing.T) {
	N := 18
	switches := make([]*p2p.Switch, N)

	// directory to store address books
	dir, err := ioutil.TempDir("", "pex_reactor")
	require.Nil(t, err)
	defer os.RemoveAll(dir) // nolint: errcheck

	books := make([]*addrBook, N)
	logger := log.TestingLogger()

	// create switches
	for i := 0; i < N; i++ {
		switches[i] = p2p.MakeSwitch(cfg, i, "testing", "123.123.123", func(i int, sw *p2p.Switch) *p2p.Switch {
			books[i] = NewAddrBook(filepath.Join(dir, fmt.Sprintf("addrbook%d.json", i)), false)
			books[i].SetLogger(logger.With("pex", i))
			sw.SetAddrBook(books[i])

			sw.SetLogger(logger.With("pex", i))

			r := NewPEXReactor(books[i], &PEXReactorConfig{})
			r.SetLogger(logger.With("pex", i))
			r.SetEnsurePeersPeriod(250 * time.Millisecond)
			sw.AddReactor("pex", r)

			return sw
		})
	}

	addOtherNodeAddrToAddrBook := func(switchIndex, otherSwitchIndex int) {
		addr := switches[otherSwitchIndex].NodeInfo().NetAddress()
		books[switchIndex].AddAddress(addr, addr)
	}

	addOtherNodeAddrToAddrBook(0, 1)
	addOtherNodeAddrToAddrBook(1, 0)
	addOtherNodeAddrToAddrBook(2, 1)
	addOtherNodeAddrToAddrBook(3, 1)
	addOtherNodeAddrToAddrBook(4, 1)
	addOtherNodeAddrToAddrBook(5, 1)
	addOtherNodeAddrToAddrBook(6, 1)
	addOtherNodeAddrToAddrBook(7, 1)
	addOtherNodeAddrToAddrBook(8, 1)
	addOtherNodeAddrToAddrBook(9, 1)
	addOtherNodeAddrToAddrBook(10, 1)
	addOtherNodeAddrToAddrBook(11, 1)
	addOtherNodeAddrToAddrBook(12, 1)
	addOtherNodeAddrToAddrBook(13, 1)
	addOtherNodeAddrToAddrBook(14, 1)
	addOtherNodeAddrToAddrBook(15, 1)
	addOtherNodeAddrToAddrBook(16, 1)
	addOtherNodeAddrToAddrBook(17, 1)

	for i, sw := range switches {
		sw.AddListener(p2p.NewDefaultListener("tcp://"+sw.NodeInfo().ListenAddr, "", false, logger.With("pex", i)))

		err := sw.Start() // start switch and reactors
		require.Nil(t, err)
	}

	assertPeersWithTimeout(t, switches, 10*time.Millisecond, 10*time.Second, N-1)

	// test at least one switch has more than 10 outbound peers
	moreThan10OutboundPeers := false
	for _, sw := range switches {
		out, _, _ := sw.NumPeers()
		if out > 10 {
			moreThan10OutboundPeers = true
		}
	}
	if !moreThan10OutboundPeers {
		t.Fatal("None of the switches has more than 10 outbound peers")
	}

	// stop them
	for _, s := range switches {
		s.Stop()
	}
}
```

I used modified version of TestPEXReactorRunning to test the expected behaviour.

File: p2p/switch.go [hunk 1]
```diff
@@ -491,11 +492,14 @@ func (sw *Switch) listenerRoutine(l Listener) {
 			break
 		}
 
-		// ignore connection if we already have enough
-		// leave room for MinNumOutboundPeers
-		maxPeers := sw.config.MaxNumPeers - DefaultMinNumOutboundPeers
```
Comment (ValarDragon) [inline p2p/switch.go]: Should we have some bias towards accepting incoming peers depending on our current inbound / outbound peer ratio? Thats something we can do postlaunch though.

Comment (melekes) [reply p2p/switch.go]: I thought about it a little bit. We can give greater control by introducing MaxNumInboundPeer/MaxNumOutboundPeers or InOutPeerRatio or something else. But it's perfectly fine to do it later.

Comment (ValarDragon) [reply p2p/switch.go]: Sounds good to me! Should we make an issue for that?

**Comment by ebuchman** [comment]
I'm not sure we should do this - see https://github.com/tendermint/tendermint/issues/2130#issuecomment-412153064

We should fix the log statement to include the min outbound peers since I agree its confusing that we don't say it there.

**Comment by melekes** [comment]
Replaced MaxNumPeers with MaxNumInboundPeers/MaxNumOutboundPeers.

File: p2p/switch.go [hunk 2]
```diff
@@ -491,11 +492,15 @@ func (sw *Switch) listenerRoutine(l Listener) {
...
-		if maxPeers <= sw.peers.Size() {
-			sw.Logger.Info("Ignoring inbound connection: already have enough peers", "address", inConn.RemoteAddr().String(), "numPeers", sw.peers.Size(), "max", maxPeers)
+		// Ignore connection if we already have enough peers.
+		_, in, _ := sw.NumPeers()
+		if in >= sw.config.MaxNumInboundPeers {
```
Comment (melekes) [inline p2p/switch.go]: note: if MaxNumInboundPeers is negative, number of conns is unlimited => which can cause consensus crash due to not enough file descriptors. Should we check and panic?

Comment (ebuchman) [reply p2p/switch.go]: We should probably catch issues in the config much sooner. Maybe we add ValidateBasic methods to the config struct we can call after parsing?

Comment (melekes) [reply p2p/switch.go]: I like this idea

**Pull request merged**